src/Compilers/CSharp/Test/Emit3/Semantics/ExtensionOperatorsTests.cs

```diff
@@ -5729,6 +5787,2123 @@ static void Main()
+        }
+    }
+}
+
+public struct S1
+{}
+
+public struct S2
+{
+    public static implicit operator S1(S2 x) => default;
+}
+""";
+            var src2 = $$$"""
+public static class Extensions2
+{
+    extension(S1)
+    {
+        public static S1 operator {{{op[0]}}}(S1 x, S1 y)
+        {
+            System.Console.Write("operator3");
+            return x;
+        }
+        public static bool operator {{{(op == "&&" ? "false" : "true")}}}(S1 x)
+        {
+            System.Console.Write("operator4");
+            return false;
+        }
+
+        public static bool operator {{{(op == "&&" ? "true" : "false")}}}(S1 x) => throw null;
+    }
+}
+""";
+
+            var comp = CreateCompilation(src1, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (28,17): error CS0217: In order to be applicable as a short circuit operator a user-defined logical operator ('Extensions1.extension(S1).operator &(S1, S1)') must have the same return type and parameter types
+                //             _ = s1 && s2;
+                Diagnostic(ErrorCode.ERR_BadBoolOp, "s1 " + op + " s2").WithArguments("NS.Extensions1.extension(S1).operator " + op[0] + "(S1, S1)").WithLocation(28, 17)
+                );
+
+            comp = CreateCompilation([src1, src2], options: TestOptions.DebugExe);
+
+            // PROTOTYPE: Should we move on to the next scope and finding Extensions2 instead of failing?
+            comp.VerifyDiagnostics(
+                // (28,17): error CS0217: In order to be applicable as a short circuit operator a user-defined logical operator ('Extensions1.extension(S1).operator &(S1, S1)') must have the same return type and parameter types
+                //             _ = s1 && s2;
+                Diagnostic(ErrorCode.ERR_BadBoolOp, "s1 " + op + " s2").WithArguments("NS.Extensions1.extension(S1).operator " + op[0] + "(S1, S1)").WithLocation(28, 17)
+                );
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void Binary_049_Consumption_Logical_TrueFalseBetterness([CombinatorialValues("&&", "||")] string op)
+        {
+            var src = $$$"""
+public static class Extensions1
+{
+    extension<T, S>((T, S))
+    {
+        public static (T, S) operator {{{op[0]}}}((T, S) x, (T, S) y)
+        {
+            System.Console.Write("operator1");
+            return x;
+        }
+
+        public static bool operator true((T, S) x) => throw null;
+        public static bool operator false((T, S) x) => throw null;
+    }
+
+    extension((int, int))
+    {
+        public static bool operator {{{(op == "&&" ? "false" : "true")}}}((int, int) x)
+        {
+            System.Console.Write("operator2");
+            return false;
+        }
+
+        public static bool operator {{{(op == "&&" ? "true" : "false")}}}((int, int) x) => throw null;
+    }
+}
+
+class Program
+{
+    static void Main()
+    {
+        var s1 = (1, 2);
+        s1 = s1 {{{op}}} s1;
```

Comment:
Consider verifying the info from semantic model in scenario with "generic operator"